core/modules/replica/IngestServerConnection.cc

```diff
@@ -0,0 +1,520 @@
+                           ptr,
+                           bytes)) return false;
+
+    // Parse the response to see what should be done next.
+
+    ptr->parse(message, bytes);
+    return true;
+}
+}   // namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+size_t IngestServerConnection::networkBufSizeBytes = 1024 * 1024;
+
+
+IngestServerConnection::Ptr IngestServerConnection::create(
+        ServiceProvider::Ptr const& serviceProvider,
+        string const& workerName,
+        boost::asio::io_service& io_service) {
+    return IngestServerConnection::Ptr(
+        new IngestServerConnection(
+            serviceProvider,
+            workerName,
+            io_service));
+}
+
+
+IngestServerConnection::IngestServerConnection(ServiceProvider::Ptr const& serviceProvider,
+                                               string const& workerName,
+                                               boost::asio::io_service& io_service)
+    :   _serviceProvider(serviceProvider),
+        _workerName(workerName),
+        _workerInfo(serviceProvider->config()->workerInfo(workerName)),
+        _socket(io_service),
+        _bufferPtr(make_shared<ProtocolBuffer>(
+            serviceProvider->config()->requestBufferSizeBytes()
+        )) {
+}
+
+
+IngestServerConnection::~IngestServerConnection() {
+    _closeFile();
+}
+
+
+void IngestServerConnection::beginProtocol() {
+    _receiveHandshake();
+}
+
+
+void IngestServerConnection::_receiveHandshake() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context << __func__);
+
+    const size_t bytes = sizeof(uint32_t);
+
+    _bufferPtr->resize(bytes);
+
+    boost::asio::async_read(
+        _socket,
+        boost::asio::buffer(_bufferPtr->data(), bytes),
+        boost::asio::transfer_at_least(bytes),
+        boost::bind(
+            &IngestServerConnection::_handshakeReceived,
+            shared_from_this(),
+            boost::asio::placeholders::error,
+            boost::asio::placeholders::bytes_transferred
+        )
+    );
+}
+
+
+void IngestServerConnection::_handshakeReceived(boost::system::error_code const& ec,
+                                                size_t bytes_transferred) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context << __func__);
+
+    if (::isErrorCode(ec, __func__)) return;
+
+    // Now read the body of the request
+
+    ProtocolIngestHandshakeRequest request;
+    if (not ::readMessage(_socket, _bufferPtr, _bufferPtr->parseLength(), request)) return;
+
+    _transactionId   = request.transaction_id();
+    _table           = request.table();
+    _chunk           = request.chunk();
+    _isOverlap       = request.is_overlap();
+    _columnSeparator = request.column_separator() == ProtocolIngestHandshakeRequest::COMMA ? ',' : '\t';
+
+    // Check if a context of the request is valid
+    
+    try {
+        auto transactionInfo = _serviceProvider->databaseServices()->transaction(_transactionId);
+        if (transactionInfo.state != "STARTED") {
```

Comment:
The transaction state is presently defined as a string in the database schema. I may migrate it to `enum` later on.

Reply:
I don't care much how it is defined in database, I'm thinking about C++ side. Even if it a string you could define constants for those strings to avoid typos.